src/main/java/org/eclipse/uprotocol/communication/InMemorySubscriber.java

```diff
@@ -66,31 +92,51 @@ public InMemorySubscriber (UTransport transport, RpcClient rpcClient) {
      * 
      * The API will return a {@link CompletionStage} with the response {@link SubscriptionResponse} or exception
      * with the failure if the subscription was not successful. The API will also register the listener to be
-     * called when messages are received.
+     * called when messages are received and allow the caller to register a listener to be notified of changes
+     * to the subscription state (ex. PENDING to SUBSCRIBED, or SUBSCRIBED to UNSUBSCRIBED, etc...).
      * 
      * @param topic The topic to subscribe to.
      * @param listener The listener to be called when a message is received on the topic.
      * @param options The call options for the subscription.
-     * @return Returns the CompletionStage with the response UMessage or exception with the failure
+     * @param handler {@link SubscriptionChangeHandler} to handle changes to subscription states.
+     * @return Returns the CompletionStage with {@link SubscriptionResponse} or exception with the failure
      * reason as {@link UStatus}.
      */
     @Override
-    public CompletionStage<SubscriptionResponse> subscribe(UUri topic, UListener listener, CallOptions options) {
+    public CompletionStage<SubscriptionResponse> subscribe(UUri topic, UListener listener, CallOptions options,
+        SubscriptionChangeHandler handler) {
         Objects.requireNonNull(topic, "Subscribe topic missing");
         Objects.requireNonNull(listener, "Request listener missing");
+        options = Objects.requireNonNullElse(options, CallOptions.DEFAULT);
 
-        final UUri subscribe = UriFactory.fromProto(
-            USubscriptionProto.getDescriptor().getServices().get(0), METHOD_SUBSCRIBE); 
-        
         final SubscriptionRequest request = SubscriptionRequest.newBuilder()
             .setTopic(topic)
             .setSubscriber(SubscriberInfo.newBuilder().setUri(transport.getSource()).build())
             .build();
         
+        // Send the subscription request and handle the response
         return RpcMapper.mapResponse(rpcClient.invokeMethod(
-                subscribe, UPayload.pack(request), options), SubscriptionResponse.class)
-            .thenApplyAsync(response -> {
-                transport.registerListener(topic, listener).toCompletableFuture().join();
+                SUBSCRIBE_METHOD, UPayload.pack(request), options), SubscriptionResponse.class)
+            
+            // Then register the listener to be called when messages are received
+            .thenCompose(response -> {
+                if ( response.getStatus().getState() == SubscriptionStatus.State.SUBSCRIBED ||
+                    response.getStatus().getState() == SubscriptionStatus.State.SUBSCRIBE_PENDING) {
+                    return transport.registerListener(topic, listener).thenApply(status -> response);
+                }
+                return CompletableFuture.completedFuture(response);
```

Comment:
Shouldn't we invoke USubscription.unsubscribe() in this case as a compensating action?

Reply:
Why would I call unsubscribe at line 127?

Reply:
not exactly line 127, but if the call to `transport.registerListener(topic, listener)` yields a failed future, then you end up in a situation where the client is (logically) subscribed, but will not have its listener being invoked. Not sure if this poses a problem but at least we should make it explicit that if the invocation of this `subscribe` method fails, this inconsistent state may be the result. IMHO the client could remedy the situation by re-trying until successful (due to the idempotence of the USubscription::subscribe operation), or by unsubscribing.